File: 20211003PracticalPythonDesignPatterns/16-visitor/visitor.py
# visitor
from __future__ import annotations
from abc import ABCMeta, abstractmethod

class Visitor:
    def visit(self, s:Shape):
        print('visited shape')

    def visit(self, d:Dot):
        print('visited dot')

class IGraphic(metaclass=ABCMeta):
    @abstractmethod
    def accept(self, v:Visitor)->None: pass

class Shape(IGraphic):
    def accept(self, v:Visitor) -> None:
        v.visit(self)

class Dot(IGraphic):
    def accept(self, v:Visitor) -> None:
        v.visit(self)

File: 20211003PracticalPythonDesignPatterns/20211122Review/03_behavioral/visitor.cpp
#include "visitor.h"
#include "graphic.h"

void Visitor::visit(const Shape& s) const
{
    std::cout << "visited shape;" << std::endl;
}

void Visitor::visit(const Dot& d) const
{
    std::cout << "visited dot;" << std::endl; 
}